src/main/java/org/folio/circulation/domain/OverdueFineCalculatorService.java

```diff
@@ -55,37 +55,47 @@ private OverdueFineCalculatorService(Clients clients) {
       new OverduePeriodCalculatorService(new CalendarRepository(clients)));
   }
 
-  public CompletableFuture<Result<CheckInProcessRecords>> calculateOverdueFine(
+  public CompletableFuture<Result<LoanAndRelatedRecords>> calculateForRenewal(
+    LoanAndRelatedRecords records) {
+
+    return createOverdueFineIfNecessary(records, Scenario.RENEWAL);
+  }
+
+  public CompletableFuture<Result<CheckInProcessRecords>> calculateForCheckIn(
     CheckInProcessRecords records) {
 
-    Loan loan = records.getLoan();
+    return createOverdueFineIfNecessary(records, Scenario.CHECKIN);
+  }
+
+  <T extends LoanRelatedRecord> CompletableFuture<Result<T>> createOverdueFineIfNecessary(
```

Comment:
Why is this generic approach needed over an approach of getting the loan in the calling code, or providing two separate methods that understand those two types?

Reply:
Because it allows the most code reuse. Could you please share, what problems you see in this approach?

Reply:
To me, this approach introduces a new interface, expects two unrelated classes to implement it and makes the function signature more complicated. It seems to introduce code rather than reuse it. I might be missing something.

I tried an approach like this earlier in mod-circulation and it turned out to be quite painful for little benefit. It may well still be in the code base.

What I don't understand, is why this couldn't be done with
* a method that takes a loan
* a method that takes a LoanAndRelatedRecords (that could be in the renewal process)
* a method that takes a CheckInProcessRecords (that could be in the check in process adapter)

Reply:
Done. I kept the two methods in the class to avoid exposing nested class `Scenario` to the outside world.